// rxjs/dist/cjs/internal/NotificationFactories.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createNotification = exports.nextNotification = exports.errorNotification = exports.COMPLETE_NOTIFICATION = void 0;
exports.COMPLETE_NOTIFICATION = (function () { return createNotification('C', undefined, undefined); })();
/**
 * Generates an error notification using the createNotification function.
 *
 * @param {Error} error - The error object to be included in the notification.
 * @returns {Notification} - The created notification object.
 * @throws {TypeError} - If the `error` parameter is not an instance of Error.
 */
function errorNotification(error) {
    return createNotification('E', undefined, error);
}
exports.errorNotification = errorNotification;
/**
 * Creates and returns the next notification with a specific value.
 *
 * @param {string} value - The value to be used in the notification.
 * @returns {Object} - An object representing the notification with type 'N' and the given value.
 */
function nextNotification(value) {
    return createNotification('N', value, undefined);
}
exports.nextNotification = nextNotification;
/**
 * Creates a notification object with the specified kind, value, and error.
 *
 * @param {string} kind - The type of the notification. This should be a string describing the nature of the notification.
 * @param {*} value - The value associated with the notification. Can be any data type depending on the context.
 * @param {Error|null} error - An Error object if there was an error, otherwise null.
 * @returns {Object} A notification object containing the kind, value, and error properties.
 *
 * @example
 * const notification = createNotification('success', 'Operation completed successfully.', null);
 * console.log(notification); // { kind: 'success', value: 'Operation completed successfully.', error: null }
 */
function createNotification(kind, value, error) {
    return {
        kind: kind,
        value: value,
        error: error,
    };
}
exports.createNotification = createNotification;
