// rxjs/dist/cjs/internal/NotificationFactories.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createNotification = exports.nextNotification = exports.errorNotification = exports.COMPLETE_NOTIFICATION = void 0;
exports.COMPLETE_NOTIFICATION = (function () { return createNotification('C', undefined, undefined); })();
/**
 * Generates an error notification.
 *
 * @param {Error} error - The error object to be included in the notification.
 * @returns {Notification} - A new notification object with type 'E' and the provided error.
 */
function errorNotification(error) {
    return createNotification('E', undefined, error);
}
exports.errorNotification = errorNotification;
/**
 * Creates a notification with the specified value and type 'N'.
 *
 * @param {string} value - The value to be included in the notification.
 * @returns {Notification} A new Notification object.
 */
function nextNotification(value) {
    return createNotification('N', value, undefined);
}
exports.nextNotification = nextNotification;
/**
 * Creates a notification object containing kind, value, and error properties.
 *
 * @param {string} kind - The type of notification (e.g., 'info', 'warning', 'error').
 * @param {*} value - The data associated with the notification.
 * @param {Error|null} [error=null] - An error object if an error occurred, otherwise null.
 * @return {{kind: string, value: *, error: Error|null}} - A notification object with the provided kind, value, and error.
 */
function createNotification(kind, value, error) {
    return {
        kind: kind,
        value: value,
        error: error,
    };
}
exports.createNotification = createNotification;
