// base64-js/index.js
'use strict'

exports.byteLength = byteLength
exports.toByteArray = toByteArray
exports.fromByteArray = fromByteArray

var lookup = []
var revLookup = []
var Arr = typeof Uint8Array !== 'undefined' ? Uint8Array : Array

var code = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/'
for (var i = 0, len = code.length; i < len; ++i) {
  lookup[i] = code[i]
  revLookup[code.charCodeAt(i)] = i
}

// Support decoding URL-safe base64 strings, as Node.js does.
// See: https://en.wikipedia.org/wiki/Base64#URL_applications
revLookup['-'.charCodeAt(0)] = 62
revLookup['_'.charCodeAt(0)] = 63

/**
 * Calculates the length of the decoded output and the number of padding characters in a base64 encoded string.
 *
 * @param {string} b64 - The base64 encoded string to analyze.
 * @returns {Array} An array containing two elements:
 *                  1. The valid length of the decoded output (number of bytes).
 *                  2. The number of padding characters at the end of the input string (number of '=' characters).
 *
 * @throws {Error} Throws an error if the input string's length is not a multiple of 4.
 *
 * Example:
 * getLens('SGVsbG8gV29ybGQh'); // Returns [12, 0]
 * getLens('SGVsbG8gV29ybGQh=='); // Returns [12, 2]
 */
function getLens (b64) {
  var len = b64.length

  if (len % 4 > 0) {
    throw new Error('Invalid string. Length must be a multiple of 4')
  }

  // Trim off extra bytes after placeholder bytes are found
  // See: https://github.com/beatgammit/base64-js/issues/42
  var validLen = b64.indexOf('=')
  if (validLen === -1) validLen = len

  var placeHoldersLen = validLen === len
    ? 0
    : 4 - (validLen % 4)

  return [validLen, placeHoldersLen]
}

// base64 is 4/3 + up to two characters of the original data
/**
 * Calculates the byte length of a Base64 encoded string.
 *
 * @param {string} b64 - The Base64 encoded string to calculate the byte length for.
 * @returns {number} The calculated byte length of the input string.
 * @throws {TypeError} If the input is not a valid Base64 string.
 */
function byteLength (b64) {
  var lens = getLens(b64)
  var validLen = lens[0]
  var placeHoldersLen = lens[1]
  return ((validLen + placeHoldersLen) * 3 / 4) - placeHoldersLen
}

/**
 * Calculates the byte length of a Base64 encoded string.
 *
 * @param {string} b64 - The Base64 encoded string.
 * @param {number} validLen - The number of valid characters in the Base64 string.
 * @param {number} placeHoldersLen - The number of padding characters in the Base64 string.
 * @returns {number} - The byte length of the decoded string.
 *
 * Example:
 * const b64 = "SGVsbG8gV29ybGQh";
 * const validLen = 13;
 * const placeHoldersLen = 0;
 * console.log(_byteLength(b64, validLen, placeHoldersLen)); // Output: 12
 */
function _byteLength (b64, validLen, placeHoldersLen) {
  return ((validLen + placeHoldersLen) * 3 / 4) - placeHoldersLen
}

function toByteArray (b64) {
  var tmp
  var lens = getLens(b64)
  var validLen = lens[0]
  var placeHoldersLen = lens[1]

  var arr = new Arr(_byteLength(b64, validLen, placeHoldersLen))

  var curByte = 0

  // if there are placeholders, only get up to the last complete 4 chars
  var len = placeHoldersLen > 0
    ? validLen - 4
    : validLen

  var i
  for (i = 0; i < len; i += 4) {
    tmp =
      (revLookup[b64.charCodeAt(i)] << 18) |
      (revLookup[b64.charCodeAt(i + 1)] << 12) |
      (revLookup[b64.charCodeAt(i + 2)] << 6) |
      revLookup[b64.charCodeAt(i + 3)]
    arr[curByte++] = (tmp >> 16) & 0xFF
    arr[curByte++] = (tmp >> 8) & 0xFF
    arr[curByte++] = tmp & 0xFF
  }

  if (placeHoldersLen === 2) {
    tmp =
      (revLookup[b64.charCodeAt(i)] << 2) |
      (revLookup[b64.charCodeAt(i + 1)] >> 4)
    arr[curByte++] = tmp & 0xFF
  }

  if (placeHoldersLen === 1) {
    tmp =
      (revLookup[b64.charCodeAt(i)] << 10) |
      (revLookup[b64.charCodeAt(i + 1)] << 4) |
      (revLookup[b64.charCodeAt(i + 2)] >> 2)
    arr[curByte++] = (tmp >> 8) & 0xFF
    arr[curByte++] = tmp & 0xFF
  }

  return arr
}

function tripletToBase64 (num) {
  return lookup[num >> 18 & 0x3F] +
    lookup[num >> 12 & 0x3F] +
    lookup[num >> 6 & 0x3F] +
    lookup[num & 0x3F]
}

/**
 * Encodes a chunk of data from a Uint8Array into Base64 format.
 *
 * @param {Uint8Array} uint8 - The input data as a Uint8Array.
 * @param {number} start - The starting index of the chunk within the array.
 * @param {number} end - The ending index (exclusive) of the chunk within the array.
 * @returns {string} - The Base64 encoded string representing the chunk.
 *
 * @example
 * const data = new Uint8Array([72, 101, 108, 108, 111]);
 * const encoded = encodeChunk(data, 0, 5);
 * console.log(encoded); // Outputs: "SGVsbG8="
 *
 * @throws {Error} - If the input parameters are invalid.
 */
function encodeChunk (uint8, start, end) {
  var tmp
  var output = []
  for (var i = start; i < end; i += 3) {
    tmp =
      ((uint8[i] << 16) & 0xFF0000) +
      ((uint8[i + 1] << 8) & 0xFF00) +
      (uint8[i + 2] & 0xFF)
    output.push(tripletToBase64(tmp))
  }
  return output.join('')
}

/**
 * Converts an array of 8-bit unsigned integers to a Base64-encoded string.
 *
 * @param {Uint8Array} uint8 - The input array of 8-bit unsigned integers.
 * @returns {string} A Base64-encoded string representing the input array.
 * @throws {TypeError} If the input is not an instance of Uint8Array.
 *
 * @example
 * const result = fromByteArray(new Uint8Array([72, 101, 108, 108, 111]));
 * console.log(result); // Output: 'SGVsbG8='
 */
function fromByteArray (uint8) {
  var tmp
  var len = uint8.length
  var extraBytes = len % 3 // if we have 1 byte left, pad 2 bytes
  var parts = []
  var maxChunkLength = 16383 // must be multiple of 3

  // go through the array every three bytes, we'll deal with trailing stuff later
  for (var i = 0, len2 = len - extraBytes; i < len2; i += maxChunkLength) {
    parts.push(encodeChunk(uint8, i, (i + maxChunkLength) > len2 ? len2 : (i + maxChunkLength)))
  }

  // pad the end with zeros, but make sure to not forget the extra bytes
  if (extraBytes === 1) {
    tmp = uint8[len - 1]
    parts.push(
      lookup[tmp >> 2] +
      lookup[(tmp << 4) & 0x3F] +
      '=='
    )
  } else if (extraBytes === 2) {
    tmp = (uint8[len - 2] << 8) + uint8[len - 1]
    parts.push(
      lookup[tmp >> 10] +
      lookup[(tmp >> 4) & 0x3F] +
      lookup[(tmp << 2) & 0x3F] +
      '='
    )
  }

  return parts.join('')
}
