// .ignored/inquirer/lib/utils/paginator.js
import colors from 'yoctocolors-cjs';

/**
 * The paginator returns a subset of the choices if the list is too long.
 */

export default class Paginator {
  /**
   * @param {import("./screen-manager")} [screen]
   * @param {{isInfinite?: boolean}} [options]
   */
  constructor(screen, options = {}) {
    const { isInfinite = true } = options;
    this.lastIndex = 0;
    this.screen = screen;
    this.isInfinite = isInfinite;
  }

  /**
   * Paginates the given output based on the current active page and page size.
   *
   * @param {string} output - The output text to be paginated.
   * @param {number} active - The current active page number (1-based index).
   * @param {number} [pageSize=7] - The number of lines per page.
   * @returns {string} - The paginated output with navigation instructions.
   *
   * @example
   * // Paginate a long text with default page size
   * const paginatedText = paginate("Line 1\nLine 2\n... \nLine N", 1);
   * console.log(paginatedText);
   *
   * @example
   * // Paginate a long text with custom page size
   * const paginatedText = paginate("Line 1\nLine 2\n... \nLine N", 1, 10);
   * console.log(paginatedText);
   */
  paginate(output, active, pageSize) {
    pageSize ||= 7;
    let lines = output.split('\n');

    if (this.screen) {
      lines = this.screen.breakLines(lines);
      active = lines
        .map((lineParts) => lineParts.length)
        .splice(0, active)
        .reduce((a, b) => a + b, 0);
      lines = lines.flat();
    }

    // Make sure there's enough lines to paginate
    if (lines.length <= pageSize) {
      return output;
    }
    const visibleLines = this.isInfinite
      ? this.getInfiniteLines(lines, active, pageSize)
      : this.getFiniteLines(lines, active, pageSize);
    this.lastIndex = active;
    return (
      visibleLines.join('\n') +
      '\n' +
      colors.dim('(Move up and down to reveal more choices)')
    );
  }

  getInfiniteLines(lines, active, pageSize) {
    if (this.pointer === undefined) {
      this.pointer = 0;
    }
    const middleOfList = Math.floor(pageSize / 2);
    // Move the pointer only when the user go down and limit it to the middle of the list
    if (
      this.pointer < middleOfList &&
      this.lastIndex < active &&
      active - this.lastIndex < pageSize
    ) {
      this.pointer = Math.min(middleOfList, this.pointer + active - this.lastIndex);
    }

    // Duplicate the lines so it give an infinite list look
    const infinite = [lines, lines, lines].flat();
    const topIndex = Math.max(0, active + lines.length - this.pointer);

    return infinite.splice(topIndex, pageSize);
  }

  /**
   * Retrieves a subset of lines from an array, ensuring it does not exceed the boundaries.
   *
   * @param {Array} lines - The array of lines to extract from.
   * @param {number} active - The index of the currently active line.
   * @param {number} pageSize - The number of lines to retrieve.
   * @returns {Array} - An array containing the subset of lines starting from the `topIndex`.
   *
   * @example
   * getFiniteLines(['a', 'b', 'c', 'd', 'e'], 2, 3);
   * // returns ['b', 'c', 'd']
   */
  getFiniteLines(lines, active, pageSize) {
    let topIndex = active - pageSize / 2;
    if (topIndex < 0) {
      topIndex = 0;
    } else if (topIndex + pageSize > lines.length) {
      topIndex = lines.length - pageSize;
    }
    return lines.splice(topIndex, pageSize);
  }
}
