// .ignored/inquirer/lib/ui/baseUI.js
import readline from 'node:readline';
import MuteStream from 'mute-stream';

/**
 * Base interface class other can inherits from
 */

export default class UI {
  constructor(opt) {
    // Instantiate the Readline interface
    // @Note: Don't reassign if already present (allow test to override the Stream)
    this.rl ||= readline.createInterface(setupReadlineOptions(opt));

    this.rl.resume();

    this.onForceClose = this.onForceClose.bind(this);

    // Make sure new prompt start on a newline when closing
    process.on('exit', this.onForceClose);

    // Terminate process on SIGINT (which will call process.on('exit') in return)
    this.rl.on('SIGINT', this.onForceClose);
  }

  /**
   * Handle the ^C exit
   * @return {null}
   */

  onForceClose() {
    this.close();
    process.kill(process.pid, 'SIGINT');
    console.log('');
  }

  /**
    // Remove events listeners
    /**
     * Handle the ^C exit.
     * @return {null} - This function does not return any value.
     */
    this.rl.removeListener('SIGINT', this.onForceClose);
    process.removeListener('exit', this.onForceClose);

    this.rl.output.unmute();

    if (this.activePrompt && typeof this.activePrompt.close === 'function') {
    // Close the readline
    /**
     * Close the interface and cleanup listeners.
     */
    this.rl.output.end();
    this.rl.pause();
    this.rl.close();
  }
}

/**
 * Sets up readline options for prompts, ensuring they are rendered correctly based on the environment.
 *
 * @param {Object} [opt={}] - The input options object. Defaults to an empty object if not provided.
 * @returns {Object} - An object containing configured readline options.
 * @throws {Error} - If prompts cannot be meaningfully rendered in a non-TTY environment and `skipTTYChecks` is false.
 *
 * @example
 * const readlineOptions = setupReadlineOptions({ skipTTYChecks: true });
 * console.log(readlineOptions);
 * // Output: { terminal: true, skipTTYChecks: true, input: [process.stdin], output: [ms] }
 */
function setupReadlineOptions(opt = {}) {
  // Inquirer 8.x:
  // opt.skipTTYChecks = opt.skipTTYChecks === undefined ? opt.input !== undefined : opt.skipTTYChecks;
  opt.skipTTYChecks = opt.skipTTYChecks === undefined ? true : opt.skipTTYChecks;

  // Default `input` to stdin
  const input = opt.input || process.stdin;

  // Check if prompt is being called in TTY environment
  // If it isn't return a failed promise
  if (!opt.skipTTYChecks && !input.isTTY) {
    const nonTtyError = new Error(
      'Prompts can not be meaningfully rendered in non-TTY environments',
    );
    nonTtyError.isTtyError = true;
    throw nonTtyError;
  }

  // Add mute capabilities to the output
  const ms = new MuteStream();
  ms.pipe(opt.output || process.stdout);
  const output = ms;

  return {
    terminal: true,
    ...opt,
    input,
    output,
  };
}
