// .ignored/inquirer/lib/ui/baseUI.js
import readline from 'node:readline';
import MuteStream from 'mute-stream';

/**
 * Base interface class other can inherits from
 */

export default class UI {
  constructor(opt) {
    // Instantiate the Readline interface
    // @Note: Don't reassign if already present (allow test to override the Stream)
    this.rl ||= readline.createInterface(setupReadlineOptions(opt));

    this.rl.resume();

    this.onForceClose = this.onForceClose.bind(this);

    // Make sure new prompt start on a newline when closing
    process.on('exit', this.onForceClose);

    // Terminate process on SIGINT (which will call process.on('exit') in return)
    this.rl.on('SIGINT', this.onForceClose);
  }

  /**
   * Handle the ^C exit
   * @return {null}
   */

  onForceClose() {
    this.close();
    process.kill(process.pid, 'SIGINT');
    console.log('');
  }

  /**
    // Remove events listeners
    /**
     * Handle the ^C exit.
     *
     * This method is designed to handle the termination of the process when the user
     * presses Ctrl+C (^C). It performs a graceful shutdown by closing any resources
     * and then terminating the process with a SIGINT signal.
     *
     * @return {null}
     */
    this.rl.removeListener('SIGINT', this.onForceClose);
    process.removeListener('exit', this.onForceClose);

    this.rl.output.unmute();

    if (this.activePrompt && typeof this.activePrompt.close === 'function') {
    // Close the readline
    /**
     * Close the interface and cleanup listeners
     *
     * @function close
     */
    this.rl.output.end();
    this.rl.pause();
    this.rl.close();
  }
}

/**
 * Sets up options for readline interface, ensuring it works correctly in TTY environments.
 *
 * @param {Object} [opt={}] - Options object to customize the behavior of the readline interface.
 * @param {boolean} [opt.skipTTYChecks=false] - If true, skips checking if the input is a TTY. Default is false.
 * @param {stream.Readable} [opt.input=process.stdin] - Input stream for the readline interface. Defaults to stdin.
 * @param {stream.Writable} [opt.output=process.stdout] - Output stream for the readline interface. Defaults to stdout.
 *
 * @returns {Object} - An object containing the configured options for the readline interface.
 *
 * @throws {Error} - Throws an error if the input is not a TTY and skipTTYChecks is false.
 */
function setupReadlineOptions(opt = {}) {
  // Inquirer 8.x:
  // opt.skipTTYChecks = opt.skipTTYChecks === undefined ? opt.input !== undefined : opt.skipTTYChecks;
  opt.skipTTYChecks = opt.skipTTYChecks === undefined ? true : opt.skipTTYChecks;

  // Default `input` to stdin
  const input = opt.input || process.stdin;

  // Check if prompt is being called in TTY environment
  // If it isn't return a failed promise
  if (!opt.skipTTYChecks && !input.isTTY) {
    const nonTtyError = new Error(
      'Prompts can not be meaningfully rendered in non-TTY environments',
    );
    nonTtyError.isTtyError = true;
    throw nonTtyError;
  }

  // Add mute capabilities to the output
  const ms = new MuteStream();
  ms.pipe(opt.output || process.stdout);
  const output = ms;

  return {
    terminal: true,
    ...opt,
    input,
    output,
  };
}
