// .ignored/commander/lib/argument.js
const { InvalidArgumentError } = require('./error.js');

class Argument {
  /**
   * Initialize a new command argument with the given name and description.
   * The default is that the argument is required, and you can explicitly
   * indicate this with <> around the name. Put [] around the name for an optional argument.
   *
   * @param {string} name
   * @param {string} [description]
   */

  constructor(name, description) {
    this.description = description || '';
    this.variadic = false;
    this.parseArg = undefined;
    this.defaultValue = undefined;
    this.defaultValueDescription = undefined;
    this.argChoices = undefined;

    switch (name[0]) {
      case '<': // e.g. <required>
        this.required = true;
        this._name = name.slice(1, -1);
        break;
      case '[': // e.g. [optional]
        this.required = false;
        this._name = name.slice(1, -1);
        break;
      default:
        this.required = true;
        this._name = name;
        break;
    }

    if (this._name.length > 3 && this._name.slice(-3) === '...') {
      this.variadic = true;
      this._name = this._name.slice(0, -3);
    }
  }


  /**
   * Returns the name of the current instance.
   *
   * @return {string} - The name associated with this instance.
   */
  name() {
    return this._name;
  }


  /**
   * @private
   * Concatenates a new value to an existing array or returns a new array containing the value.
   *
   * @param {*} value - The new value to be concatenated.
   * @param {Array|*} previous - The previous value, which can either be an array or any other type.
   * @returns {Array} - An array containing the concatenated values.
   * @throws {TypeError} - If `previous` is not an array and not equal to `defaultValue`.
   */
  _concatValue(value, previous) {
    if (previous === this.defaultValue || !Array.isArray(previous)) {
      return [value];
    }

    return previous.concat(value);
  }


  /**
   * Set the default value, and optionally supply the description to be displayed in the help.
   *
   * @param {*} value - The default value to be set.
   * @param {string} [description] - Optional description to be displayed in the help.
   * @returns {Argument} - Returns the current instance of Argument for method chaining.
   */
  default(value, description) {
    this.defaultValue = value;
    this.defaultValueDescription = description;
    return this;
  }


  /**
   * Set the custom handler for processing CLI command arguments into argument values.
   *
   * @param {Function} [fn] - The function to be called for parsing command arguments. If not provided, any previously set handler will be removed.
   * @return {Argument} - Returns the current instance of `Argument` to allow method chaining.
   */
  argParser(fn) {
    this.parseArg = fn;
    return this;
  }


  /**
   * Sets the allowed argument values.
   *
   * @param {string[]} values - An array of strings representing the allowable argument values.
   * @return {Argument} - Returns the current instance for method chaining.
   * @throws {InvalidArgumentError} - Throws an error if the provided argument is not in the list of choices.
   */
  choices(values) {
    this.argChoices = values.slice();
    this.parseArg = (arg, previous) => {
      if (!this.argChoices.includes(arg)) {
        throw new InvalidArgumentError(`Allowed choices are ${this.argChoices.join(', ')}.`);
      }
      if (this.variadic) {
        return this._concatValue(arg, previous);
      }
      return arg;
    };
    return this;
  }

  /**
   * Marks an argument as required.
   *
   * @return {Object} - The current object with the `required` property set to true.
   */
  argRequired() {
    this.required = true;
    return this;
  }

  /**
   * Make argument optional.
   *
   * @return {Object} The current object instance, allowing method chaining.
   */
  argOptional() {
    this.required = false;
    return this;
  }
}


/**
 * Takes an argument and returns its human readable equivalent for help usage.
 *
 * @param {Argument} arg - The argument to convert to a human-readable format.
 * @returns {string} - The human-readable representation of the argument.
 * @api private
 */
function humanReadableArgName(arg) {
  const nameOutput = arg.name() + (arg.variadic === true ? '...' : '');

  return arg.required
    ? '<' + nameOutput + '>'
    : '[' + nameOutput + ']';
}

exports.Argument = Argument;
exports.humanReadableArgName = humanReadableArgName;
