// .ignored/commander/lib/argument.js
const { InvalidArgumentError } = require('./error.js');

class Argument {
  /**
   * Initialize a new command argument with the given name and description.
   * The default is that the argument is required, and you can explicitly
   * indicate this with <> around the name. Put [] around the name for an optional argument.
   *
   * @param {string} name
   * @param {string} [description]
   */

  constructor(name, description) {
    this.description = description || '';
    this.variadic = false;
    this.parseArg = undefined;
    this.defaultValue = undefined;
    this.defaultValueDescription = undefined;
    this.argChoices = undefined;

    switch (name[0]) {
      case '<': // e.g. <required>
        this.required = true;
        this._name = name.slice(1, -1);
        break;
      case '[': // e.g. [optional]
        this.required = false;
        this._name = name.slice(1, -1);
        break;
      default:
        this.required = true;
        this._name = name;
        break;
    }

    if (this._name.length > 3 && this._name.slice(-3) === '...') {
      this.variadic = true;
      this._name = this._name.slice(0, -3);
    }
  }


  /**
   * Retrieve the argument name.
   *
   * @return {string} The name of the argument.
   */
  name() {
    return this._name;
  }


  /**
   * Concatenates a value to an existing array or returns a new array with the value.
   *
   * @private
   * @param {any} value - The value to be concatenated.
   * @param {Array|any} previous - The current array or value to which the value should be added.
   * @returns {Array} A new array containing the original values and the new value.
   *
   * @example
   * // Example usage is not provided in the code snippet.
   */
  _concatValue(value, previous) {
    if (previous === this.defaultValue || !Array.isArray(previous)) {
      return [value];
    }

    return previous.concat(value);
  }


  /**
   * Set the default value, and optionally supply the description to be displayed in the help.
   *
   * @param {*} value - The default value for the argument.
   * @param {string} [description] - Optional. The description that will be shown in the help message.
   * @return {Argument} - The current instance of Argument, allowing method chaining.
   */
  default(value, description) {
    this.defaultValue = value;
    this.defaultValueDescription = description;
    return this;
  }


  /**
   * Set the custom handler for processing CLI command arguments into argument values.
   *
   * @param {Function} [fn] - The function to use as the custom handler. If not provided, any existing handler will be removed.
   * @return {Argument} - The current instance of Argument, allowing method chaining.
   */
  argParser(fn) {
    this.parseArg = fn;
    return this;
  }


  /**
   * Set allowed choices for an argument.
   *
   * @param {string[]} values - An array of strings representing the allowed choices.
   * @return {Argument} - The current instance of the Argument class, allowing method chaining.
   *
   * @throws {InvalidArgumentError} If the provided argument is not one of the allowed choices.
   */
  choices(values) {
    this.argChoices = values.slice();
    this.parseArg = (arg, previous) => {
      if (!this.argChoices.includes(arg)) {
        throw new InvalidArgumentError(`Allowed choices are ${this.argChoices.join(', ')}.`);
      }
      if (this.variadic) {
        return this._concatValue(arg, previous);
      }
      return arg;
    };
    return this;
  }

  /**
   * Makes an argument required.
   *
   * @returns {object} - The current object instance with the 'required' property set to true.
   */
  argRequired() {
    this.required = true;
    return this;
  }

  /**
   * Make argument optional.
   *
   * @returns {object} - Returns the current object for method chaining.
   */
  argOptional() {
    this.required = false;
    return this;
  }
}


/**
 * Takes an argument and returns its human readable equivalent for help usage.
 *
 * @param {Argument} arg - The argument object to convert to a human-readable format.
 * @returns {string} - The human-readable representation of the argument, formatted as required by the command-line tool.
 * @api private
 */
function humanReadableArgName(arg) {
  const nameOutput = arg.name() + (arg.variadic === true ? '...' : '');

  return arg.required
    ? '<' + nameOutput + '>'
    : '[' + nameOutput + ']';
}

exports.Argument = Argument;
exports.humanReadableArgName = humanReadableArgName;
