// inquirer/lib/utils/paginator.js
import colors from 'yoctocolors-cjs';

/**
 * The paginator returns a subset of the choices if the list is too long.
 */

export default class Paginator {
  /**
   * @param {import("./screen-manager")} [screen]
   * @param {{isInfinite?: boolean}} [options]
   */
  constructor(screen, options = {}) {
    const { isInfinite = true } = options;
    this.lastIndex = 0;
    this.screen = screen;
    this.isInfinite = isInfinite;
  }

  /**
   * Paginates the output based on the active line and page size.
   *
   * @param {string} output - The text to be paginated.
   * @param {number} active - The index of the currently active line.
   * @param {number} [pageSize=7] - The number of lines to display per page. Default is 7.
   * @returns {string} - The paginated text with navigation instructions.
   *
   * @example
   * // Paginates a string output with default pageSize
   * paginate("line1\nline2\nline3", 0);
   * // Returns: "line1\nline2\nline3\n(Move up and down to reveal more choices)"
   *
   * @example
   * // Paginates a string output with custom pageSize
   * paginate("line1\nline2\nline3\nline4\nline5", 0, 5);
   * // Returns: "line1\nline2\nline3\n(Move up and down to reveal more choices)"
   */
  paginate(output, active, pageSize) {
    pageSize ||= 7;
    let lines = output.split('\n');

    if (this.screen) {
      lines = this.screen.breakLines(lines);
      active = lines
        .map((lineParts) => lineParts.length)
        .splice(0, active)
        .reduce((a, b) => a + b, 0);
      lines = lines.flat();
    }

    // Make sure there's enough lines to paginate
    if (lines.length <= pageSize) {
      return output;
    }
    const visibleLines = this.isInfinite
      ? this.getInfiniteLines(lines, active, pageSize)
      : this.getFiniteLines(lines, active, pageSize);
    this.lastIndex = active;
    return (
      visibleLines.join('\n') +
      '\n' +
      colors.dim('(Move up and down to reveal more choices)')
    );
  }

  /**
   * Generates an infinite list of lines by duplicating the input list.
   *
   * @param {Array} lines - The array of lines to be duplicated and used for creating an infinite list.
   * @param {number} active - The index of the currently active item in the list.
   * @param {number} pageSize - The number of items to return per page.
   * @returns {Array} - An array containing the selected page from the infinite list.
   *
   * @throws {TypeError} if `lines` is not an array or if `active` and `pageSize` are not numbers.
   */
  getInfiniteLines(lines, active, pageSize) {
    if (this.pointer === undefined) {
      this.pointer = 0;
    }
    const middleOfList = Math.floor(pageSize / 2);
    // Move the pointer only when the user go down and limit it to the middle of the list
    if (
      this.pointer < middleOfList &&
      this.lastIndex < active &&
      active - this.lastIndex < pageSize
    ) {
      this.pointer = Math.min(middleOfList, this.pointer + active - this.lastIndex);
    }

    // Duplicate the lines so it give an infinite list look
    const infinite = [lines, lines, lines].flat();
    const topIndex = Math.max(0, active + lines.length - this.pointer);

    return infinite.splice(topIndex, pageSize);
  }

  /**
   * Retrieves a finite number of lines from an array based on the current active index and page size.
   *
   * @param {Array} lines - The array of lines to extract from.
   * @param {number} active - The currently active line index.
   * @param {number} pageSize - The number of lines to retrieve per page.
   * @returns {Array} An array containing the finite number of lines based on the provided parameters.
   *
   * @example
   * const lines = ['line1', 'line2', 'line3', 'line4', 'line5'];
   * const result = getFiniteLines(lines, 2, 3);
   * console.log(result); // Output: ['line2', 'line3', 'line4']
   *
   * @note This function ensures that the top index does not go out of bounds by clamping it between 0 and the array's length minus the page size.
   */
  getFiniteLines(lines, active, pageSize) {
    let topIndex = active - pageSize / 2;
    if (topIndex < 0) {
      topIndex = 0;
    } else if (topIndex + pageSize > lines.length) {
      topIndex = lines.length - pageSize;
    }
    return lines.splice(topIndex, pageSize);
  }
}
