// readable-stream/errors-browser.js
'use strict';

/**
 * Helper function to enable loose inheritance between two classes.
 *
 * @private
 * @param {Function} subClass - The subclass that extends the superclass.
 * @param {Function} superClass - The superclass being extended.
 */
function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

var codes = {};

/**
 * Creates a new error type with a specified code, message, and optional base class.
 *
 * @param {string} code - The unique identifier for the error type.
 * @param {(string|Function)} message - The error message or a function that returns the error message. If a function is provided, it should accept up to three arguments (arg1, arg2, arg3).
 * @param {Error} [Base=Error] - The base class for the new error type. Defaults to `Error` if not specified.
 *
 * @throws {TypeError} Throws an error if `code` is not a string or `message` is neither a string nor a function.
 */
function createErrorType(code, message, Base) {
  if (!Base) {
    Base = Error;
  }

  /**
   * Retrieves a message based on the provided arguments.
   *
   * @param {string|Function} message - A string message or a function that returns a message.
   * @param {any} arg1 - The first argument passed to the function if 'message' is a function.
   * @param {any} arg2 - The second argument passed to the function if 'message' is a function.
   * @param {any} arg3 - The third argument passed to the function if 'message' is a function.
   * @returns {string} - The retrieved message.
   * @throws {Error} - If 'message' is neither a string nor a function that returns a string.
   *
   * @example
   * // Using a string message
   * const result1 = getMessage("Hello, world!");
   * console.log(result1); // Output: Hello, world!
   *
   * @example
   * // Using a function to generate the message
   * const messageFunction = (name) => `Hello, ${name}!`;
   * const result2 = getMessage(messageFunction, "John");
   * console.log(result2); // Output: Hello, John!
   */
  function getMessage(arg1, arg2, arg3) {
    if (typeof message === 'string') {
      return message;
    } else {
      return message(arg1, arg2, arg3);
    }
  }

  var NodeError =
  /*#__PURE__*/
  function (_Base) {
    _inheritsLoose(NodeError, _Base);

    /**
     * Constructs an instance of NodeError with a custom message based on provided arguments.
     *
     * @constructor
     * @extends _Base
     *
     * @param {string} arg1 - The first argument used to generate the error message.
     * @param {string} arg2 - The second argument used to generate the error message.
     * @param {string} arg3 - The third argument used to generate the error message.
     *
     * @returns {NodeError} A new instance of NodeError.
     */
    function NodeError(arg1, arg2, arg3) {
      return _Base.call(this, getMessage(arg1, arg2, arg3)) || this;
    }

    return NodeError;
  }(Base);

  NodeError.prototype.name = Base.name;
  NodeError.prototype.code = code;
  codes[code] = NodeError;
} // https://github.com/nodejs/node/blob/v10.8.0/lib/internal/errors.js


function oneOf(expected, thing) {
  if (Array.isArray(expected)) {
    var len = expected.length;
    expected = expected.map(function (i) {
      return String(i);
    });

    if (len > 2) {
      return "one of ".concat(thing, " ").concat(expected.slice(0, len - 1).join(', '), ", or ") + expected[len - 1];
    } else if (len === 2) {
      return "one of ".concat(thing, " ").concat(expected[0], " or ").concat(expected[1]);
    } else {
      return "of ".concat(thing, " ").concat(expected[0]);
    }
  } else {
    return "of ".concat(thing, " ").concat(String(expected));
  }
} // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/startsWith


/**
 * Determines whether a string begins with the specified value.
 *
 * @param {string} str - The string to search.
 * @param {string} search - The substring to search for at the beginning of the string.
 * @param {number} [pos=0] - Optional. The position in the string at which to begin searching. If omitted, 0 is used.
 * @returns {boolean} Returns true if the string begins with the specified value; otherwise, false.
 *
 * @example
 * // Returns true
 * startsWith('Hello, world!', 'Hello');
 *
 * @example
 * // Returns false
 * startsWith('Hello, world!', 'world', 7);
 */
function startsWith(str, search, pos) {
  return str.substr(!pos || pos < 0 ? 0 : +pos, search.length) === search;
} // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/endsWith


/**
 * Determines whether the string ends with the specified suffix.
 *
 * @param {string} str - The string to check.
 * @param {string} search - The substring to search for at the end of the string.
 * @param {number} [this_len=undefined] - Optional. If provided, it is used as the length of the string to consider.
 * @returns {boolean} - Returns true if the string ends with the specified suffix; otherwise, false.
 *
 * @example
 * // returns true
 * endsWith("Hello World", "World");
 *
 * @example
 * // returns false
 * endsWith("Hello World", "Universe");
 */
function endsWith(str, search, this_len) {
  if (this_len === undefined || this_len > str.length) {
    this_len = str.length;
  }

  return str.substring(this_len - search.length, this_len) === search;
} // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/includes


/**
 * Checks if a string contains a specified substring within a range of characters.
 *
 * @param {string} str - The string to search in.
 * @param {string} search - The substring to search for.
 * @param {number} [start=0] - The index at which to start the search.
 * @returns {boolean} - True if the substring is found, false otherwise.
 *
 * @example
 * includes("Hello world", "world");
 * // returns true
 *
 * @example
 * includes("Hello world", "world", 7);
 * // returns true
 *
 * @example
 * includes("Hello world", "world", 8);
 * // returns false
 */
function includes(str, search, start) {
  if (typeof start !== 'number') {
    start = 0;
  }

  if (start + search.length > str.length) {
    return false;
  } else {
    return str.indexOf(search, start) !== -1;
  }
}

createErrorType('ERR_INVALID_OPT_VALUE', function (name, value) {
  return 'The value "' + value + '" is invalid for option "' + name + '"';
}, TypeError);
createErrorType('ERR_INVALID_ARG_TYPE', function (name, expected, actual) {
  // determiner: 'must be' or 'must not be'
  var determiner;

  if (typeof expected === 'string' && startsWith(expected, 'not ')) {
    determiner = 'must not be';
    expected = expected.replace(/^not /, '');
  } else {
    determiner = 'must be';
  }

  var msg;

  if (endsWith(name, ' argument')) {
    // For cases like 'first argument'
    msg = "The ".concat(name, " ").concat(determiner, " ").concat(oneOf(expected, 'type'));
  } else {
    var type = includes(name, '.') ? 'property' : 'argument';
    msg = "The \"".concat(name, "\" ").concat(type, " ").concat(determiner, " ").concat(oneOf(expected, 'type'));
  }

  msg += ". Received type ".concat(typeof actual);
  return msg;
}, TypeError);
createErrorType('ERR_STREAM_PUSH_AFTER_EOF', 'stream.push() after EOF');
createErrorType('ERR_METHOD_NOT_IMPLEMENTED', function (name) {
  return 'The ' + name + ' method is not implemented';
});
createErrorType('ERR_STREAM_PREMATURE_CLOSE', 'Premature close');
createErrorType('ERR_STREAM_DESTROYED', function (name) {
  return 'Cannot call ' + name + ' after a stream was destroyed';
});
createErrorType('ERR_MULTIPLE_CALLBACK', 'Callback called multiple times');
createErrorType('ERR_STREAM_CANNOT_PIPE', 'Cannot pipe, not readable');
createErrorType('ERR_STREAM_WRITE_AFTER_END', 'write after end');
createErrorType('ERR_STREAM_NULL_VALUES', 'May not write null values to stream', TypeError);
createErrorType('ERR_UNKNOWN_ENCODING', function (arg) {
  return 'Unknown encoding: ' + arg;
}, TypeError);
createErrorType('ERR_STREAM_UNSHIFT_AFTER_END_EVENT', 'stream.unshift() after end event');
module.exports.codes = codes;
